File: leetcode/lc-0067/solution.cpp
class Solution {
public:
    string addBinary(string a, string b) {
        if (a == "0") return b;
        if (b == "0") return a;

        // 111 + 111 = 1011 => 4 digits
        int m = a.size();
        int n = b.size();
        int k = max(m, n) + 1;
        string result(k, '0');

        for (int i = m - 1, j = k - 1; i >= 0; i--, j--) {
            result[j] += (a[i] - '0');
        }
        for (int i = n - 1, j = k - 1; i >= 0; i--, j--) {
            result[j] += (b[i] - '0');
        }
        for (int i = k - 1; i > 0; i--) {
            int sum = result[i] - '0';
            result[i] = sum % 2 + '0';
            result[i - 1] += sum / 2;
        }

        for (int i = 0; i < k; i++) {
            if (result[i] != '0') return result.substr(i);
        }
        return result;
    }
};

File: leetcode/lc-0067/solution.rs
impl Solution {
    pub fn add_binary(a: String, b: String) -> String {
        if a == "0" {
            return b;
        }
        if b == "0" {
            return a;
        }

        // 111 + 111 = 1011 => 4 digits
        let m = a.len();
        let n = b.len();
        let k = std::cmp::max(m, n) + 1;
        let mut result = vec![b'0'; k];

        let mut j = k - 1;
        let a = a.as_bytes();
        for i in (0..m).rev() {
            result[j] += (a[i] - b'0');
            j -= 1;
        }
        let mut j = k - 1;
        let b = b.as_bytes();
        for i in (0..n).rev() {
            result[j] += (b[i] - b'0');
            j -= 1;
        }
        for i in (1..k).rev() {
            let sum = result[i] - b'0';
            result[i] = (sum as i32 % 2) as u8 + b'0';
            result[i - 1] += (sum / 2) as u8;
        }

        for i in 0..k {
            if result[i] != b'0' {
                return String::from_utf8(result[i..].to_vec()).unwrap();
            }
        }
        String::from_utf8(result).unwrap()
    }
}
